src/geometry.jl

```diff
@@ -81,6 +82,130 @@ end
 
 @compat public revolution_volume
 push!(document[Symbol("Geometry")], :revolution_volume)
+"""
+    r_intersect_interval(x0::T, y0::T, dx::T, dy::T, r_min::T, r_max::T) where {T<:Real}
+
+Finds intersection of ray starting at x0 and y0 with direction (dx,dy) 
+    with reference major radius value r_ref
+
+- `x0``, `y0``: Origin
+- `dx``, `dy``: Normalized direction in x-y plane
+- `r_min`, `r_max`: Edges of r_range
+
+Returns interval [t1, t2] for which r_min <= r(t) <= r_max with r(t)^2 = (x+dx*t)^2 + (y+dy*t)^2
+
+"""
+
+function solve_r_intersect(x0::T, y0::T, dx::T, dy::T, r_min::T, r_max::T) where {T<:Real}
+    r0 = sqrt(x0^2 + y0^2)
+    t_crossings = zeros(Float64, 4)
+    crossing = zeros(Float64, 4) # +1 -> into domain
+                                 # -1 -> out of 
+                                 # 0 -> no crossing
+    into_domain = [1.0, -1.0]
+    if dx == 0 && dy == 0
+        # Handle vertical ray
+        if (r0 > r_min && r0 < r_max)
+            return [[0.0 Inf];]
+        else
+            return zeros(Float64, 0)
+        end
+    end
+    for (i_ref, r_ref) in enumerate([r_min, r_max])
+        Δ = r_ref^2 * (dx^2 + dy^2) - x0^2 * dy^2 - y0^2 * dx^2 + 2 * x0 * y0* dx*dy
+        if Δ < 0
+            # println("No intersection for ", r_ref)
+            t_crossings[2*(i_ref-1) + 1] = Inf
+            t_crossings[2*(i_ref-1) + 2] = Inf
+            crossing[2*(i_ref-1) + 1] = 0.0
+            crossing[2*(i_ref-1) + 2] = 0.0
+            continue
+        end
+        t1 = -(x0*dx + y0 * dy) - sqrt(Δ)
+        t2 = -(x0*dx + y0 * dy) + sqrt(Δ)
+        t_crossings[2*(i_ref-1) + 1] = t1 >= 0 ? t1/ (dx^2 + dy^2) : Inf
+        t_crossings[2*(i_ref-1) + 2] = t2 >= 0 ? t2/ (dx^2 + dy^2) : Inf
+        crossing[2*(i_ref-1) + 1] = t1 >= 0 ? into_domain[i_ref]*sign(x0*dx + dx^2*t1 + y0 * dy + dy^2*t1) : 0.0
+        crossing[2*(i_ref-1) + 2] = t2 >= 0 ? into_domain[i_ref]*sign(x0*dx + dx^2*t2 + y0 * dy + dy^2*t2) : 0.0
+    end
+    last = 0 # 1 in, 0 out
+    # If the first point is in we add it to the crossings
+    if (r0 > r_min && r0 < r_max)
+        append!(t_crossings, 0.0)
+        append!(crossing, 1.0)
+    end
+    i_sort = sortperm(t_crossings)
+    intervals = zeros(Float64, 0)
+    for i in i_sort
+        if t_crossings[i] == Inf
+            continue
+        end
+        if last == 0 && crossing[i] > 0.0 # looking for a crossing into the rectangle
+            append!(intervals, t_crossings[i])
+            last = 1
+        elseif last == 1 && crossing[i] < 0.0 # looking for a crossing leaving the rectangle
+            append!(intervals, t_crossings[i])
+            last = 0
+        end
+    end
+    if length(intervals) % 2 != 0
+        throw(ErrorException("Somehow only found odd number of intersections which is impossible."))
```

Comment:
Make sure grazing intersections are properly handled.

Reply:
For my specific use-case grazing intersection can raise an error. A beam that hits the rectangular flux matrix parallel to one of its boundaries indicates an error in how the beam is set up.